File: backend/app/collectors/github_collector.py
from datetime import datetime

import requests

from app.config import GITHUB_TOKEN

API_BASE = "https://api.github.com"


def fetch_recent_runs(full_name: str, per_page: int = 10) -> list[dict]:
    headers = {"Accept": "application/vnd.github+json"}
    if GITHUB_TOKEN:
        headers["Authorization"] = f"Bearer {GITHUB_TOKEN}"

    resp = requests.get(
        f"{API_BASE}/repos/{full_name}/actions/runs",
        params={"per_page": per_page},
        headers=headers,
        timeout=15,
    )
    resp.raise_for_status()

    runs = []
    for run in resp.json().get("workflow_runs", []):
        started_at = datetime.strptime(run["run_started_at"], "%Y-%m-%dT%H:%M:%SZ")
        updated_at = datetime.strptime(run["updated_at"], "%Y-%m-%dT%H:%M:%SZ")
        duration_s = int((updated_at - started_at).total_seconds())

        runs.append({
            "run_id": run["id"],
            "workflow_name": run.get("name") or run.get("workflow_id"),
            "status": run["status"],
            "conclusion": run.get("conclusion"),
            "started_at": started_at,
            "duration_s": max(duration_s, 0),
            "html_url": run["html_url"],
        })
    return runs

File: backend/src/collectors/githubCollector.ts
import { githubToken } from "../config";

export interface DeploymentRunResult {
  run_id: number;
  workflow_name: string;
  status: string;
  conclusion: string | null;
  started_at: Date;
  duration_s: number;
  html_url: string;
}

const API_BASE = "https://api.github.com";

interface GithubWorkflowRun {
  id: number;
  name: string | null;
  workflow_id: number;
  status: string;
  conclusion: string | null;
  run_started_at: string;
  updated_at: string;
  html_url: string;
}

export async function fetchRecentRuns(fullName: string, perPage = 10): Promise<DeploymentRunResult[]> {
  const headers: Record<string, string> = { Accept: "application/vnd.github+json" };
  if (githubToken) headers.Authorization = `Bearer ${githubToken}`;

  const response = await fetch(
    `${API_BASE}/repos/${fullName}/actions/runs?per_page=${perPage}`,
    { headers },
  );
  if (!response.ok) {
    throw new Error(`GitHub API error ${response.status}: ${await response.text()}`);
  }

  const body = (await response.json()) as { workflow_runs: GithubWorkflowRun[] };

  return body.workflow_runs.map((run) => {
    const startedAt = new Date(run.run_started_at);
    const updatedAt = new Date(run.updated_at);
    const durationS = Math.max(
      Math.round((updatedAt.getTime() - startedAt.getTime()) / 1000),
      0,
    );

    return {
      run_id: run.id,
      workflow_name: run.name ?? String(run.workflow_id),
      status: run.status,
      conclusion: run.conclusion,
      started_at: startedAt,
      duration_s: durationS,
      html_url: run.html_url,
    };
  });
}
